src/rust/logger/src/single.rs

```diff
@@ -0,0 +1,178 @@
+
+use crate::*;
+use std::io::{Error, Write};
+
+/// Implements a basic logger which sends all log messages to a single queue.
+pub(crate) struct Logger {
+    log_filled: Queue<LogBuffer>,
+    log_cleared: Queue<LogBuffer>,
+    buffer_size: usize,
+    format: FormatFunction,
+    level_filter: LevelFilter,
+}
+
+impl Logger {
+    pub fn level_filter(&self) -> LevelFilter {
+        self.level_filter
+    }
+}
+
+impl Log for Logger {
+    fn enabled(&self, metadata: &log::Metadata<'_>) -> bool {
+        metadata.level() <= self.level_filter
+    }
+
+    fn log(&self, record: &log::Record<'_>) {
+        // If the log message is filtered by the log level, return early.
+        if !self.enabled(record.metadata()) {
+            return;
+        }
+
+        // Tries to re-use a buffer from the pool or allocate a new buffer to
+        // to avoid blocking and try to avoid dropping the message. Message may
+        // still be dropped if the log_filled queue is full.
+        let mut buffer = self
+            .log_cleared
+            .pop()
+            .unwrap_or_else(|| Vec::with_capacity(self.buffer_size));
+
+        // Write the log message into the buffer and send to the receiver
+        if (self.format)(&mut buffer, recent_local(), record).is_ok() {
+            // Note this may drop a log message, but avoids blocking. The
+            // preference here is to preserve log messages which lead up to the
+            // point where we begin to drop log messages. For example, if an
+            // error begins to happen which causes very many log messages, it is
+            // more beneficial to have the history leading up to the issue than
+            // to preserve more recent error messages.
+            let _ = self.log_filled.push(buffer);
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+/// Implements a basic drain type which receives log messages over a queue and
+/// flushes them to a single buffered output.
+pub(crate) struct LogDrain {
+    log_filled: Queue<LogBuffer>,
+    log_cleared: Queue<LogBuffer>,
+    buffer_size: usize,
+    output: Box<dyn Output>,
+}
+
+impl Drain for LogDrain {
+    fn flush(&mut self) -> Result<(), Error> {
+        while let Some(mut log_buffer) = self.log_filled.pop() {
+            let _ = self.output.write(&log_buffer);
+
+            // shrink oversized buffer
+            if log_buffer.len() > self.buffer_size {
+                log_buffer.truncate(self.buffer_size);
+                log_buffer.shrink_to_fit();
+                log_buffer.clear();
+            }
+
+            // recycle the buffer, buffer will be dropped if the pool is full
+            log_buffer.clear();
+            let _ = self.log_cleared.push(log_buffer);
+        }
+        self.output.flush()
+    }
+}
+
+/// A type to construct a basic `AsyncLog` which routes all log messages to a
+/// single `Output`.
+pub struct LogBuilder {
+    total_buffer_size: usize,
+    log_message_size: usize,
```

Comment:
Given the structure of log buffer, would it be better to expose `log_message_size` (or named more explicitly as `single_message_size`) and `log_queue_depth` instead of doing the division based on size?

I think the memory consideration of log buffer has been too small to matter in production, so providing a number-of-message config knob may be more user-friendly.